### DISCUSSION TITLE
Better « How to write shaders » guide

### DISCUSSION THREAD

**Discussion by hadronized** [Rust-GPU/rust-gpu#191] [General]
Hello!

I’m looking for a way to write my shaders using `rust-gpu`, and I realized it’s not as easy as I thought it would be. The main issue I have is the guide that doesn’t really go deep in details on how to setup the various crates, and the toolchain.

From my understanding, given what `cargo-gpu` does, it should be possible to compile the shader codes without impacting the toolchain of the main project, even though it depends on the SPIR-V builder.

Also, I saw there was a way to bypass the toolchain check, but I haven’t been successful at it so far (it keeps complaining about the wrong toolchain being used).

Thanks in advance!

**Comment by schell** [comment]
Hi @hadronized :) 

I haven't started on the `cargo-gpu` manual yet, but if you use the version that's on `main` as of this reply, you *should* be able to run `cargo gpu build --shader-crate path/to/your/crate` and it *should* determine and install the necessary toolchain for you, as well as compiling the `rustc` backend that does SPIR-V codegen and invoking `cargo` with the necessary args. 

If you run into any trouble please open a ticket at the [cargo-gpu repo](https://github.com/rust-gpu/cargo-gpu).

**Comment by hadronized** [comment]
Cool, thanks for the fast reply. I think I struggle the most with is the crate separation that is required because of all the toolchain setup. I’m not sure whether it’s the tooling that’s hard to work with, or whether it’s just me not used to it.

I’ll play along with it, thanks!

**Reply by schell** [reply]
Oh, you don't have to separate your crate into GPU and CPU crates! You can combine them, but when compilingfor GPU the crate must be `no_std`. I use this annotation in my own project: `#![cfg_attr(target_arch = "spirv", no_std)]` (along with a bunch of others). This will allow you to have CPU code in your shader crate, so long as it is feature/target_arch gated. You can peruse my project at <https://github.com/schell/renderling> if you'd like to see how I set this up.

**Reply by hadronized** [reply]
Ah that’s cool! I’m trying out to invoke `cargo-gpu` in my `build.rs`, is there any recommended way to do that besides just using `Command::new()`?

**Reply by schell** [reply]
In my own project I have building shaders as an explicit step that I have to invoke on the command line, because compile times can be quite long, and I often don't want to compile my shaders every time a file changes.

But if you _do_ want to do that, I think `Command::new()` (or something similar) is probably your best bet. `cargo-gpu` is not set up to be used as a library. At least not yet.

**Comment by git-staus** [comment]
> Also, I saw there was a way to bypass the toolchain check, but I haven’t been successful at it so far (it keeps complaining about the wrong toolchain being used).

For future readers, I (believe) this is the way to do that.
```
RUSTGPU_SKIP_TOOLCHAIN_CHECK=1 cargo build
```
When I did this it happily obliged and started producing `rustc_codegen_spirv` compilation errors, so maybe don't do this. (I think I ran into #366)

**Comment by Firestar99** [comment]
I started on a new user-guide a few weeks back, but haven't gotten any further than a single page due to getting sick. Anyway, here's the page: https://github.com/Rust-GPU/rust-gpu/blob/new_docs/docs/src/user-guide/crate.md

@schell I want to include cargo-gpu and the direct fixed toolchain path in the guide. I'm also considering giving cargo-gpu a final v2 refactor before finishing it, to clean up some stuff that may break things but allows future features to be easier to implement, like `cargo clippy`. More on that to come.

**Reply by schell** [reply]
Ok, let me know how I can help 👍 .

**Reply by git-staus** [reply]
I have been looking to get into `rust-gpu` recently but lost my momentum somewhat due to technical roadblocks. If feedback from a novice would make the process of writing the guide easier I would be happy help out at any stage in the process.

**Reply by Firestar99** [reply]
Thanks, I'll let you know when it's ready for a proper review :D

**Reply by alphastrata** [reply]
I'm keen to lend a hand if required see #473.

I think it'd be worth adding as I've done in those templates throwing in in a :puke-emoji `vscode` settings json so folk can point things at the right build targets -- we're gonna get inundated by ppl complaining that RA isn't telling them anything... (maybe?)
